dipy/denoise/nlmeans.py

```diff
@@ -18,83 +13,219 @@ def nlmeans(
+    See :footcite:p:`Coupe2008` and :footcite:p:`Coupe2012` for further details
+    about the blockwise method.
 
     Parameters
     ----------
     arr : 3D or 4D ndarray
-        The array to be denoised
-    mask : 3D ndarray
-    sigma : float or 3D array
-        standard deviation of the noise estimated from the data
-    patch_radius : int
-        patch size is ``2 x patch_radius + 1``. Default is 1.
-    block_radius : int
-        block size is ``2 x block_radius + 1``. Default is 5.
-    rician : boolean
-        If True the noise is estimated as Rician, otherwise Gaussian noise
-        is assumed.
+        The array to be denoised. For 3D arrays, shape should be (height, width, depth).
+        For 4D arrays, shape should be (height, width, depth, volumes) where the last
+        dimension represents different volumes (e.g., DWI directions).
+    sigma : float or 1D ndarray
+        Standard deviation of the noise estimated from the data. For 3D arrays,
+        this should be a scalar. For 4D arrays, this can be either a scalar (same noise
+        level for all volumes) or a 1D array with length equal to the number of volumes.
+    mask : 3D ndarray, optional
+        Binary mask indicating which voxels to process. Should have shape
+        (height, width, depth). Voxels with mask value 0 are set to 0 in output.
+        If None, all voxels are processed.
+    patch_radius : int, optional
+        Radius for similarity search neighborhoods. Patches of size
+        (2*patch_radius + 1)³ around each voxel are compared to find similar
+        structures.
+    block_radius : int, optional
+        Radius for weighted averaging blocks. Each block has size
+        (2*block_radius + 1)³. Larger blocks provide more smoothing but
+        increase computational cost. If None, defaults are:
+
+        - method='classic': 5 (11*11*11 blocks)
+        - method='blockwise': 2 (5*5*5 blocks)
+    rician : bool, optional
+        If True, assumes Rician noise model (appropriate for magnitude MRI data).
+        If False, assumes Gaussian noise model.
     num_threads : int, optional
-        Number of threads to be used for OpenMP parallelization. If None
-        (default) the value of OMP_NUM_THREADS environment variable is used
-        if it is set, otherwise all available threads are used. If < 0 the
-        maximal number of threads minus $|num_threads + 1|$ is used (enter -1 to
-        use as many threads as possible). 0 raises an error.
+        Number of OpenMP threads to use for parallel processing. If None,
+        uses all available CPU threads. Set to 1 to disable parallel processing.
+    method : str, optional
+        Algorithm method to use:
+
+        - 'blockwise': Improved algorithm with better coordinate handling,
+          memory efficiency, and statistical pre-filtering
+        - 'classic': Original algorithm with traditional implementation
+
+        .. versionadded:: 1.12.0
```

Comment:
To be removed?

Reply:
No, actually, we need to use this more...